lib/reporting/reports/sales_tax/sales_tax_totals_by_producer.rb

```diff
@@ -0,0 +1,169 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ClassLength
+module Reporting
+  module Reports
+    module SalesTax
+      class SalesTaxTotalsByProducer < Base
+        def search
+          report_line_items.orders
+        end
+
+        def order_permissions
+          return @order_permissions unless @order_permissions.nil?
+
+          @order_permissions = ::Permissions::Order.new(user, ransack_params)
+        end
+
+        def report_line_items
+          @report_line_items ||= Reporting::LineItems.new(order_permissions, params)
+        end
+
+        def query_result
+          # The objective is to group the line items by
+          # [tax_rate, supplier_id, distributor_id and order_cycle_id]
+          # As every line item can match more than one tax_rate
+          # We'll do it in two steps
+          # 1. we group by [[tax_rate], supplier_id, distributor_id , order_cycle_id]
+          # 2. we split each key that has more than tax_rate
+          report_line_items.list.group_by do |line_item|
+            [
+              line_item.tax_rates.map(&:id),
+              line_item.supplier_id,
+              line_item.order.distributor_id,
+              line_item.order.order_cycle_id
+            ]
+          end.flat_map do |k, v|
+            k.first.map do |tax_rate_id|
+              {
+                [tax_rate_id] + k[1..] => v
+              }
+            end
+          end
```

Comment:
Thank you for the explanation. This is really hard to read.

I'm not sure if the result is really better but here's an idea which could improve the speed of the report and simplify this logic. We could join the line items with their tax rates so that we can one line for each tax rate. It means that we don't have to split the keys later on and we can simply group by tax_rate_id to start with.

Reply:
When you group in the database you get only one result per group and that group is not a line item any more. We don't actually need the line items, do we?

A problem is probably Ransack which applies `where` conditions which would need to be `having` conditions.

Reply:
>  We don't actually need the line items, do we?

We need access to the line items because every line item can match a different set of tax rates and can belong to a different supplier.

I'm not sure if this answers your question, please let me know.

Reply:
I was hoping that the database can group those line items for us. But now I think that it probably won't work with Ransack. We shouldn't waste too much time on this. It was just an idea. Basically, I was hoping for a query that returns the report table with all the data and we don't have to do any further calculation. While I think that it's possible, it may simply be too complicated to code and to maintain. It's probably best if we leave that for now and revisit if performance becomes a problem.